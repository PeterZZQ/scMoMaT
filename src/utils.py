import numpy as np
import torch
# from torch_sparse import SparseTensor
import matplotlib.pyplot as plt


import numpy as np
from scipy import stats

import torch.nn.functional as F

def lsi_ATAC(X, k = 100, use_first = False):
    """\
    Description:
    ------------
        Compute LSI with TF-IDF transform, i.e. SVD on document matrix, can do tsne on the reduced dimension

    Parameters:
    ------------
        X: cell by feature(region) count matrix
        k: number of latent dimensions
        use_first: since we know that the first LSI dimension is related to sequencing depth, we just ignore the first dimension since, and only pass the 2nd dimension and onwards for t-SNE
    
    Returns:
    -----------
        latent: cell latent matrix
    """    
    from sklearn.feature_extraction.text import TfidfTransformer
    from sklearn.decomposition import TruncatedSVD

    # binarize the scATAC-Seq count matrix
    bin_X = np.where(X < 1, 0, 1)
    
    # perform Latent Semantic Indexing Analysis
    # get TF-IDF matrix
    tfidf = TfidfTransformer(norm='l2', sublinear_tf=True)
    normed_count = tfidf.fit_transform(bin_X)

    # perform SVD on the sparse matrix
    lsi = TruncatedSVD(n_components = k, random_state=42)
    lsi_r = lsi.fit_transform(normed_count)
    
    # use the first component or not
    if use_first:
        return lsi_r
    else:
        return lsi_r[:, 1:]

def kl_div(C):
    C_uni = torch.ones_like(C)
    C_uni = C_uni/torch.sum(C_uni, dim = 1)[:,None]
    return torch.median(torch.sum(C * torch.log(C/C_uni), dim = 1))
    


def quantile_norm(X):
    """Normalize the columns of X to each have the same distribution.

    Given an expression matrix (microarray data, read counts, etc) of M genes
    by N samples, quantile normalization ensures all samples have the same
    spread of data (by construction).

    The data across each row are averaged to obtain an average column. Each
    column quantile is replaced with the corresponding quantile of the average
    column.

    Parameters
    ----------
    X : 2D array of float, shape (M, N)
        The input data, with M rows (genes/features) and N columns (samples).

    Returns
    -------
    Xn : 2D array of float, shape (M, N)
        The normalized data.
    """
    # compute the quantiles
    quantiles = np.mean(np.sort(X, axis=0), axis=1)

    # compute the column-wise ranks. Each observation is replaced with its
    # rank in that column: the smallest observation is replaced by 1, the
    # second-smallest by 2, ..., and the largest by M, the number of rows.
    ranks = np.apply_along_axis(stats.rankdata, 0, X)

    # convert ranks to integer indices from 0 to M-1
    rank_indices = ranks.astype(int) - 1

    # index the quantiles for each rank with the ranks matrix
    Xn = quantiles[rank_indices]

    return(Xn)

def quantile_norm_log(X):
    logX = np.log1p(X)
    logXn = quantile_norm(logX)
    return logXn

def preprocess(counts, mode = "standard", modality = "RNA"):
    """\
    Description:
    ------------
    Preprocess the dataset
    """
    if mode == "standard":
        if modality == "ATAC":
            counts = (counts > 0).astype(np.float)
            
        # normalize according to the library size
        libsize = np.median(np.sum(counts, axis = 1))
        counts = counts / np.sum(counts, axis = 1)[:,None] * libsize

        if modality == "RNA":
            counts = np.log1p(counts)

    elif mode == "quantile":
        if modality == "RNA":
            counts = quantile_norm_log(counts)
        else:
            # make binary
            counts = (counts > 0).astype(np.float) 
            # counts = counts / np.sum(counts, axis = 1)[:,None]
    
    elif mode == "gact":
        # gene by region matrix
        counts = counts/(np.sum(counts, axis = 1)[:,None] + 1e-6)


    return counts

def plot_latent(z1, z2, anno1 = None, anno2 = None, mode = "joint", save = None, figsize = (20,10), axis_label = "Latent", **kwargs):
    """\
    Description
        Plot latent space
    Parameters
        z1
            the latent space of first data batch, of the shape (n_samples, n_dimensions)
        z2
            the latent space of the second data batch, of the shape (n_samples, n_dimensions)
        anno1
            the cluster annotation of the first data batch, of the  shape (n_samples,)
        anno2
            the cluster annotation of the second data batch, of the  shape (n_samples,)
        mode
            "joint": plot two latent spaces(from two batches) into one figure
            "separate" plot two latent spaces separately
        save
            file name for the figure
        figsize
            figure size
    """
    _kwargs = {
        "s": 10,
        "alpha": 0.9,
    }
    _kwargs.update(kwargs)

    fig = plt.figure(figsize = figsize)
    if mode == "modality":
        colormap = plt.cm.get_cmap("tab10")
        ax = fig.add_subplot()
        ax.scatter(z1[:,0], z1[:,1], color = colormap(1), label = "scRNA-Seq", **_kwargs)
        ax.scatter(z2[:,0], z2[:,1], color = colormap(2), label = "scATAC-Seq", **_kwargs)
        ax.legend(loc='upper left', prop={'size': 15}, frameon = False, ncol = 1, bbox_to_anchor=(1.04, 1))
        
        ax.tick_params(axis = "both", which = "major", labelsize = 15)

        ax.set_xlabel(axis_label + " 1", fontsize = 19)
        ax.set_ylabel(axis_label + " 2", fontsize = 19)
        ax.spines['right'].set_visible(False)
        ax.spines['top'].set_visible(False)  

    elif mode == "joint":
        ax = fig.add_subplot()
        cluster_types = set([x for x in np.unique(anno1)]).union(set([x for x in np.unique(anno2)]))
        colormap = plt.cm.get_cmap("tab20", len(cluster_types))

        for i, cluster_type in enumerate(cluster_types):
            index = np.where(anno1 == cluster_type)[0]
            index2 = np.where(anno2 == cluster_type)[0]
            ax.scatter(np.concatenate((z1[index,0], z2[index2,0])), np.concatenate((z1[index,1],z2[index2,1])), color = colormap(i), label = cluster_type, **_kwargs)
        
        ax.legend(loc='upper left', prop={'size': 15}, frameon = False, ncol = 1, bbox_to_anchor=(1.04, 1))
        
        ax.tick_params(axis = "both", which = "major", labelsize = 15)

        ax.set_xlabel(axis_label + " 1", fontsize = 19)
        ax.set_ylabel(axis_label + " 2", fontsize = 19)
        ax.spines['right'].set_visible(False)
        ax.spines['top'].set_visible(False)  


    elif mode == "separate":
        axs = fig.subplots(1,2)
        cluster_types = set([x for x in np.unique(anno1)]).union(set([x for x in np.unique(anno2)]))
        colormap = plt.cm.get_cmap("tab20", len(cluster_types))

        
        for i, cluster_type in enumerate(cluster_types):
            index = np.where(anno1 == cluster_type)[0]

            if index.shape[0] != 0:
                axs[0].scatter(z1[index,0], z1[index,1], color = colormap(i), label = cluster_type, **_kwargs)
        # axs[0].legend(fontsize = font_size)
        # axs[0].legend(loc='upper left', prop={'size': 15}, frameon = False, ncol = 1, bbox_to_anchor=(1.04, 1))
        axs[0].set_title("scRNA-Seq", fontsize = 25)

        axs[0].tick_params(axis = "both", which = "major", labelsize = 15)

        axs[0].set_xlabel(axis_label + " 1", fontsize = 19)
        axs[0].set_ylabel(axis_label + " 2", fontsize = 19)
        axs[0].set_xlim(np.min(np.concatenate((z1[:,0], z2[:,0]))), np.max(np.concatenate((z1[:,0], z2[:,0]))))
        axs[0].set_ylim(np.min(np.concatenate((z1[:,1], z2[:,1]))), np.max(np.concatenate((z1[:,1], z2[:,1]))))
        axs[0].spines['right'].set_visible(False)
        axs[0].spines['top'].set_visible(False)  


        for i, cluster_type in enumerate(cluster_types):
            index = np.where(anno2 == cluster_type)[0]

            if index.shape[0] != 0:
                axs[1].scatter(z2[index,0], z2[index,1], color = colormap(i), label = cluster_type, **_kwargs)
        # axs[1].axis("off")
        axs[1].legend(loc='upper left', prop={'size': 15}, frameon = False, ncol = 1, bbox_to_anchor=(1.04, 1))
        axs[1].set_title("scATAC-Seq", fontsize = 25)

        axs[1].tick_params(axis = "both", which = "major", labelsize = 15)

        axs[1].set_xlabel(axis_label + " 1", fontsize = 19)
        axs[1].set_ylabel(axis_label + " 2", fontsize = 19)
        axs[1].set_xlim(np.min(np.concatenate((z1[:,0], z2[:,0]))), np.max(np.concatenate((z1[:,0], z2[:,0]))))
        axs[1].set_ylim(np.min(np.concatenate((z1[:,1], z2[:,1]))), np.max(np.concatenate((z1[:,1], z2[:,1]))))
        axs[1].spines['right'].set_visible(False)
        axs[1].spines['top'].set_visible(False)
        
    
    elif mode == "hybrid":
        axs = fig.subplots(1,2)
        cluster_types = set([x for x in np.unique(anno1)]).union(set([x for x in np.unique(anno2)]))
        colormap = plt.cm.get_cmap("tab20", len(cluster_types))

        for i, cluster_type in enumerate(cluster_types):
            index = np.where(anno1 == cluster_type)[0]
            index2 = np.where(anno2 == cluster_type)[0]
            axs[1].scatter(np.concatenate((z1[index,0], z2[index2,0])), np.concatenate((z1[index,1],z2[index2,1])), color = colormap(i), label = cluster_type, **_kwargs)
        
        axs[1].legend(loc='upper left', prop={'size': 15}, frameon = False, ncol = 1, bbox_to_anchor=(1.04, 1), markerscale=4)
        
        axs[1].tick_params(axis = "both", which = "major", labelsize = 15)

        axs[1].set_xlabel(axis_label + " 1", fontsize = 19)
        axs[1].set_ylabel(axis_label + " 2", fontsize = 19)
        axs[1].spines['right'].set_visible(False)
        axs[1].spines['top'].set_visible(False)          

        colormap = plt.cm.get_cmap("tab10")
        axs[0].scatter(z1[:,0], z1[:,1], color = colormap(1), label = "scRNA-Seq", **_kwargs)
        axs[0].scatter(z2[:,0], z2[:,1], color = colormap(2), label = "scATAC-Seq", **_kwargs)
        axs[0].legend(loc='upper left', prop={'size': 15}, frameon = False, ncol = 1, bbox_to_anchor=(0.54, 1), markerscale=4)
        
        axs[0].tick_params(axis = "both", which = "major", labelsize = 15)

        axs[0].set_xlabel(axis_label + " 1", fontsize = 19)
        axs[0].set_ylabel(axis_label + " 2", fontsize = 19)
        axs[0].spines['right'].set_visible(False)
        axs[0].spines['top'].set_visible(False) 
        
        
    if save:
        fig.savefig(save, bbox_inches = "tight")
    

# def csr2st(A):
#     A = A.tocoo()
#     col = torch.LongTensor(A.col)
#     row = torch.LongTensor(A.row)
#     value = torch.FloatTensor(A.data)
#     sparse_sizes = A.shape
#     return SparseTensor(row=row, col=col, value=value, sparse_sizes=sparse_sizes)


def _pairwise_distances(x, y = None):
    x_norm = (x**2).sum(1).view(-1, 1)
    # calculate the pairwise distance between two datasets
    if y is not None:
        y_t = torch.transpose(y, 0, 1)
        y_norm = (y**2).sum(1).view(1, -1)
    else:
        y_t = torch.transpose(x, 0, 1)
        y_norm = x_norm.view(1, -1)
    
    dist = x_norm + y_norm - 2.0 * torch.mm(x, y_t)
    # Ensure diagonal is zero if x=y
    if y is None:
        dist = dist - torch.diag(dist.diag)

    return torch.clamp(dist, 0.0, np.inf)

def match_alignment(z_rna, z_atac, k = 10):
    # note that the distance is squared version
    dist = _pairwise_distances(z_atac, z_rna).numpy()
    knn_index = np.argpartition(dist, kth = k - 1, axis = 1)[:,(k-1)]
    kth_dist = np.take_along_axis(dist, knn_index[:,None], axis = 1)
    
    K = dist/(kth_dist + 1e-6) 
    K = (dist <= kth_dist) * np.exp(-K) 
    K = K/np.sum(K, axis = 1)[:,None]

    z_atac = torch.FloatTensor(K).mm(z_rna)
    return z_rna, z_atac


def match_clust(z_rna, z_atac, k = 10, scale = 1):
    # note that the distance is squared version
    clust_rna = np.argmax(z_rna, axis = 1).squeeze()
    clust_atac = np.argmax(z_atac, axis = 1).squeeze()
    for clust in range(np.max((z_rna.shape[1], z_atac.shape[1]))):
        rna_idx = np.where(clust_rna == clust)[0]
        atac_idx = np.where(clust_atac == clust)[0]
        
        if (rna_idx.shape[0] != 0) and (atac_idx.shape[0] != 0):
            z_rna_clust = z_rna[rna_idx,:]
            z_atac_clust = z_atac[atac_idx,:]

            
            dist = _pairwise_distances(z_atac_clust, z_rna_clust).numpy()
            knn_index = np.argpartition(dist, kth = k - 1, axis = 1)[:,(k-1)]
            kth_dist = np.take_along_axis(dist, knn_index[:,None], axis = 1)

            K = dist/(kth_dist * scale)
            K = (dist <= kth_dist) * np.exp(-K) 
            K = K/np.sum(K, axis = 1)[:,None]

            z_atac[atac_idx,:] = torch.FloatTensor(K).mm(z_rna_clust)
        else:
            print("no cell in cluster {:d}".format(clust))
    return z_rna, z_atac    



###############################################################################


def compute_pairwise_distances(x, y):
    x_norm = (x**2).sum(1).view(-1, 1)
    y_t = torch.transpose(y, 0, 1)
    y_norm = (y**2).sum(1).view(1, -1)
    
    dist = x_norm + y_norm - 2.0 * torch.mm(x, y_t)
    return torch.clamp(dist, 0.0, np.inf)


def _gaussian_kernel_matrix(x, y):
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    sigmas = torch.FloatTensor([1e-6, 1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1, 5, 10, 15, 20, 25, 30, 35, 100, 1e3, 1e4, 1e5, 1e6]).to(device)
    dist = compute_pairwise_distances(x, y)
    beta = 1. / (2. * sigmas[:,None])
    s = - beta.mm(dist.reshape((1, -1)) )
    result =  torch.sum(torch.exp(s), dim = 0)
    return result


def maximum_mean_discrepancy(x, y): #Function to calculate MMD value
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    cost = torch.mean(_gaussian_kernel_matrix(x, x))
    cost += torch.mean(_gaussian_kernel_matrix(y, y))
    cost -= 2.0 * torch.mean(_gaussian_kernel_matrix(x, y))
    cost = torch.sqrt(cost ** 2 + 1e-9)
    if cost.data.item()<0:
        cost = torch.FloatTensor([0.0]).to(device)

    return cost