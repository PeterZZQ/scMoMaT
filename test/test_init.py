import sys, os
sys.path.append('../')
sys.path.append('../src/')

import torch
import numpy as np
from torch.nn import Module, Parameter
from utils import preprocess
import torch.nn.functional as F

from sklearn.decomposition import PCA
from umap import UMAP

import pandas as pd 
import numpy as np 
import scipy.sparse as sp

from sklearn.metrics import adjusted_rand_score
import seaborn as sns
import matplotlib.pyplot as plt

import newmodel
import utils

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")


#######################################################################

# Test on simulated dataset

#######################################################################


def batch_mixing_entropy(C1, C2, runs = 1):
    # construct knn graph
    k = 100
    
    C = torch.cat((C1, C2), dim = 0)
    dist = utils._pairwise_distances(C, C).numpy()
    knn_index = np.argpartition(dist, kth = k - 1, axis = 1)[:,(k-1)]
    kth_dist = np.take_along_axis(dist, knn_index[:,None], axis = 1)
    knn = ((dist - kth_dist[:, None]) <= 0)


    # select random anchors, and calculate the entropy
    entropys = []
    for run in range(runs):
        random_anchor = np.random.choice(C1.shape[0] + C2.shape[0], 100, replace=False)
        p1 = np.sum(knn[random_anchor,:C1.shape[0]], axis = 1)/100
        p2 = np.sum(knn[random_anchor,C1.shape[0]:], axis = 1)/100
        entropys.append(np.sum(p1 * np.log(p1 + 1e-6)) + np.sum(p2 * np.log(p2 + 1e-6)))
    return np.array(entropys)
    
    
# Read in the data: 2 batches, 3 clusters
dir = '../data/simulated/'

paths = ['2b5c_ziqi1/', '2b5c_ziqi2/']

for path in paths:
    for init in ["SVD", "NMF", "random"]:
        latent_dim = 5
        batchsize = 0.5
        
        best_loss = 1e12
        kl1 = []
        kl2 = []
        
        losses = []
        ari_scores = []
        for run in range(1,5):
            print("data: "+ path[:-1] + "\ninit: "+ init + 
                  "\nbatch_size: "+ str(batchsize) + "\nrun: "+ str(run))
            
            # read in the data
            counts_rna = pd.read_csv(os.path.join(dir + path, 'GxC1.txt'), sep = "\t", header = None).values.T
            counts_atac = pd.read_csv(os.path.join(dir + path, 'RxC2.txt'), sep = "\t", header = None).values.T
            A = pd.read_csv(os.path.join(dir + path, 'region2gene.txt'), sep = "\t", header = None).values.T

            counts_rna = np.array(counts_rna)
            counts_atac = np.array(counts_atac)
            A = np.array(A)
            counts = {"rna":[counts_rna], "atac": [counts_atac], "gact": [A]}
            
            # init
            if init == "SVD":
                model_init = newmodel.cfrm_svdinit(counts, N = latent_dim).to(device)
                C_1, C_2, A_g, A_r, C_g, C_r = model_init.train_func(match = True)
                init_val = [C_1, C_2, A_g, A_r, C_g, C_r]
            elif init == "NMF":
                model1 = newmodel.cfrm_nmfinit(counts, N = 5, batch_size = 0.5, lr = 1e-3, seed = 0).to(device)
                C_1, C_2, A_g, A_r, C_g, C_r = model1.train_func(T = 500, match = True)
                init_val = [C_1, C_2, A_g, A_r, C_g, C_r]
            else:
                init_val = None

            # train model
            model1 = newmodel.cfrm_new(counts, N = latent_dim, batch_size = batchsize, interval=20000, lr=1e-3, 
                                       alpha = [1000, 1000, 100, 100], seed = run, init = init_val).to(device)

            model1.train_func(T = 10000)

            # validation
            with torch.no_grad():
                loss, *_ = model1.batch_loss('valid')
                print('Final Loss is {:.5f}'.format(loss.item()))
                print()

            # save the best of all runs
            if loss < best_loss:
                torch.save(model1.state_dict(), "./results/models/" + path[:-1] + "_" + init + "_" + "_best.pt")

            # plots
            umap_op = UMAP(n_components = 2, n_neighbors = 15, min_dist = 0.4) 

            z_rna = model1.softmax(model1.C_1.cpu().detach()).numpy()
            z_atac = model1.softmax(model1.C_2.cpu().detach()).numpy()


            max_rna = np.argmax(z_rna, axis = 1).squeeze()
            max_atac = np.argmax(z_atac, axis = 1).squeeze()
            z = umap_op.fit_transform(np.concatenate((z_rna, z_atac), axis = 0))

            label_rna = pd.read_csv(os.path.join(dir + path, "cell_label1.txt"), index_col=0, sep = "\t")["pop"].values.squeeze()
            label_atac = pd.read_csv(os.path.join(dir + path, "cell_label2.txt"), index_col=0, sep = "\t")["pop"].values.squeeze()
            utils.plot_latent(z[:z_rna.shape[0],:], z[z_rna.shape[0]:,:], label_rna, label_atac, mode= "separate", save = "./results/plots/" + path[:-1] + "_" + init + "_" + str(run) + ".png")

            # kl divergence
            kl1.append(utils.kl_div(model1.softmax(model1.C_1)).item())
            kl2.append(utils.kl_div(model1.softmax(model1.C_2)).item())
            losses.append(loss.item())

            # benchmark how??
            # batch mixing entropy
            bm_entro = batch_mixing_entropy(model1.softmax(model1.C_1.cpu().detach()), 
                                            model1.softmax(model1.C_2.cpu().detach()), runs = 50)
            np.save( "./results/" + path[:-1] + "_" + init + "_" + str(run) + "_bm_enropy.npy", np.array(bm_entro))
            fig = plt.figure(figsize = (5,5))
            ax = fig.add_subplot()
            sns.boxplot(x = pd.DataFrame(data = bm_entro[:,None], columns = ["bm_entro"])["bm_entro"], ax = ax)
            fig.savefig("./results/"+ path[:-1] + "_" + init + "_" + str(run) + "_bm_enropy.png")


            # ARI score, same as what is used in Fabian benchmark code
            ari_scores.append(adjusted_rand_score(labels_pred = np.concatenate((max_rna, max_atac), axis = 0), labels_true = np.concatenate((label_rna, label_atac), axis = 0)))
        np.save( "./results/" + path[:-1] + "_" + init + "_ari.npy", np.array(ari_scores))
        np.save( "./results/" + path[:-1] + "_" + init + "_loss.npy", np.array(losses))
            # alignment score (in liger), knn graph, local structure metric (in seurat v3)



        kl = np.array(kl1.append(kl2))
            
        # save kl divergence
        np.save("./results/" + path[:-1] + "_" + init + "_kl.npy", kl)
            

