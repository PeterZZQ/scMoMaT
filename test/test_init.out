---------------------------------------
Begin PBS Prologue Sun May  9 23:06:42 EDT 2021
Job ID:     6745114.sched-hive.pace.gatech.edu
User ID:    zzhang834
Job name:   test_cfrm
Queue:      hive-gpu-short
End PBS Prologue Sun May  9 23:06:42 EDT 2021
---------------------------------------
../src/utils.py:121: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  fig = plt.figure(figsize = figsize)
data: 2b5c_ziqi1
init: SVD
batch_size: 0.5
run: 1
Final Loss is -136.35779

data: 2b5c_ziqi1
init: SVD
batch_size: 0.5
run: 2
Final Loss is -136.45107

data: 2b5c_ziqi1
init: SVD
batch_size: 0.5
run: 3
Final Loss is -136.38458

data: 2b5c_ziqi1
init: SVD
batch_size: 0.5
run: 4
Final Loss is -136.46735

data: 2b5c_ziqi1
init: NMF
batch_size: 0.5
run: 1
iteration: 0, loss1: 0.059398550540208817, loss2: 0.07714065164327621
Final Loss is -142.00198

data: 2b5c_ziqi1
init: NMF
batch_size: 0.5
run: 2
iteration: 0, loss1: 0.059398550540208817, loss2: 0.07714065164327621
Final Loss is -142.00174

data: 2b5c_ziqi1
init: NMF
batch_size: 0.5
run: 3
iteration: 0, loss1: 0.059398550540208817, loss2: 0.07714065164327621
Final Loss is -142.00237

data: 2b5c_ziqi1
init: NMF
batch_size: 0.5
run: 4
iteration: 0, loss1: 0.059398550540208817, loss2: 0.07714065164327621
Final Loss is -142.00189

data: 2b5c_ziqi1
init: random
batch_size: 0.5
run: 1
Final Loss is -141.17401

data: 2b5c_ziqi1
init: random
batch_size: 0.5
run: 2
Final Loss is -142.23087

data: 2b5c_ziqi1
init: random
batch_size: 0.5
run: 3
Final Loss is -142.92210

data: 2b5c_ziqi1
init: random
batch_size: 0.5
run: 4
Final Loss is -142.49022

data: 2b5c_ziqi2
init: SVD
batch_size: 0.5
run: 1
Final Loss is -136.98424

data: 2b5c_ziqi2
init: SVD
batch_size: 0.5
run: 2
Final Loss is -137.14511

data: 2b5c_ziqi2
init: SVD
batch_size: 0.5
run: 3
Final Loss is -136.95952

data: 2b5c_ziqi2
init: SVD
batch_size: 0.5
run: 4
Final Loss is -136.94882

data: 2b5c_ziqi2
init: NMF
batch_size: 0.5
run: 1
iteration: 0, loss1: 0.08018159866333008, loss2: 0.08038871735334396
Final Loss is -145.68916

data: 2b5c_ziqi2
init: NMF
batch_size: 0.5
run: 2
iteration: 0, loss1: 0.08018159866333008, loss2: 0.08038871735334396
Final Loss is -145.69910

data: 2b5c_ziqi2
init: NMF
batch_size: 0.5
run: 3
iteration: 0, loss1: 0.08018159866333008, loss2: 0.08038871735334396
Final Loss is -145.70068

data: 2b5c_ziqi2
init: NMF
batch_size: 0.5
run: 4
iteration: 0, loss1: 0.08018159866333008, loss2: 0.08038871735334396
Final Loss is -145.70151

data: 2b5c_ziqi2
init: random
batch_size: 0.5
run: 1
Final Loss is -144.58600

data: 2b5c_ziqi2
init: random
batch_size: 0.5
run: 2
Final Loss is -144.62457

data: 2b5c_ziqi2
init: random
batch_size: 0.5
run: 3
Final Loss is -146.32278

data: 2b5c_ziqi2
init: random
batch_size: 0.5
run: 4
Final Loss is -144.72409

---------------------------------------
Begin PBS Epilogue Mon May 10 00:53:00 EDT 2021
Job ID:     6745114.sched-hive.pace.gatech.edu
User ID:    zzhang834
Job name:   test_cfrm
Resources:  nodes=1:ppn=4:gpus=1,walltime=10:20:00,neednodes=1:ppn=4:gpus=1
Rsrc Used:  cput=01:50:39,vmem=144624316kb,walltime=01:46:18,mem=134012852kb,energy_used=0
Queue:      hive-gpu-short
Nodes:     
atl1-1-01-018-31.pace.gatech.edu
End PBS Epilogue Mon May 10 00:53:00 EDT 2021
---------------------------------------
